291~295
========================

291.In ServletContext.getNamedDispatcher(string arg0) arg0 is the

A. url of a servlet relative to the ServletRequest

B. url of a servlet relative to the Context

C. name of a servlet

D. absolute url to any servlet in any Context

<!--sec data-title="解析" data-id="section291_2" data-collapse=true ces-->
答案：C

備註：根據規格書，getNamedDispatcher方法的參數是一個能被ServletContext找到的Servlet名稱，如果找得到該Servlet就包裝成RequestDispatcher物件並回傳，找不到則回傳null

相關說明：Servlet 3.0規格書9.1 Obtaining a RequestDispatcher
<!--endsec-->

---
292.When using Servlet asynchronous API if you want to dispatch the request back to a particular url -"/url" within the same Servlet Context which of the following API would you use?

A. AsyncContext.dispatch();

B. AsyncContext.dispatch("/url");

C. AsyncContext.dispatch(servletContext, "/url");

D. AsyncContext.start(runnable);

E. RequestDispatcher.fotward("/url");

F. RequestDispatcher.forward(servletContext, "/url");

G. RequestDispatcher.include("/url");

<!--sec data-title="解析" data-id="section292_2" data-collapse=true ces-->
答案：C

備註：待補
<!--endsec-->

---
293.Which describes a trigger that causes a call to an AsyncListener?

A. Completion of processing

B. Receipt of a new request from the same client

C. Addition of a new filter to the processing stream for this request

D. Completion of each timer interval

<!--sec data-title="解析" data-id="section293_2" data-collapse=true ces-->
答案：A

備註：

根據API，AsyncListener有四個方法

1. public void onComplete(AsyncEvent event)─用來通知監聽器非同步操作完成了
2. public void onError(AsyncEvent event)─用來通知監聽器非同步操作未能完成
3. public void onStartAsync(AsyncEvent event)─用來通知監聽器正在透過呼叫一個ServletRequest.startAsync方法啟動一個新的非同步周期
4. public void onTimeout(AsyncEvent event)─用來通知監聽器非同步操作已逾時

上述四個狀況發生時，其對應的AsyncContext可以透過AsyncEvent的getAsyncContext方法取得

相關說明：Servlet 3.0規格書2.3.3.3 Asynchronous processing
<!--endsec-->

---
294.Given the JavaBean class:

```java
public class MyBean {
	private Boolean roasted = false;
	public MyBean () {}
	public Boolean isRoasted () { 
		return roasted; 
	}
	public void setRoasted (Boolean roasted) { 
		this.roasted = roasted; 
	}
}
```

Assume a controller has created an instance of this bean, called setRoasted (true), and inserted the bean into
the application scope using the name "myBean".

A JSP contains these two tags:

```jsp
<jsp:useBean id = "aBean" class = "MyBean" scope = "page" />
<jsp:getProperty name = "aBean" property = "roasted" />
```

Which is true?

A. the page will include the output false

B. the page will include the output

C. the page will report that the property roasted does not exist

D. the page will report that the syntax of the useBean tag is incorrect

E. the page will report that the syntax of the getProperty tag ls incorrect

<!--sec data-title="解析" data-id="section294_2" data-collapse=true ces-->
答案：E

備註：因為getProperty會去找bean當中有沒有getXxx方法，但roasted屬性只有isRoasted方法，因此會出現例外錯誤

![1507161750141](https://github.com/Carrie-Lai/Test/blob/master/media/11115.jpeg)

改為getRoasted後則會出現另一個例外錯誤訊息：MyBean cannot be resolved to a type

將MyBean放入套件後並將useBean的class加上套件後才正常顯示false(因Controller創的物件是被放在application scope，useBean在page scope中拿不到就會創一個新的，roasted為預設值false)
<!--endsec-->

---
295.A method call inside your servlet has thrown an IOException resulting from a lost network connection to a necessary resource. The servlet has determined that it will be unable to operate successfully fen few minutes,but expects to be able to resume operation after that period. How should this situation handled?

A. The servlet should re-throw the exception to the web container.

B. The servlet should throw a ServletException to the web container.

C. The servlet should throw an UnavailableException to the web container.

D. The servlet should redirect to an error page that reports the exception message.

E. The servlet should retry the request until the method returns successfully.

<!--sec data-title="解析" data-id="section295_2" data-collapse=true ces-->
答案：C

備註：因為一些原因造成Servlet暫時無法成功處理請求，並要在一段時間後重新嘗試，可以拋出UnavailableException

相關說明：

[Servlet--ServletException类，UnavailableException类](http://blog.csdn.net/u011794238/article/details/46502471)

Servlet 3.0規格書2.3.3.2 Exceptions During Request Handling
<!--endsec-->